### PR TITLE
Refactory upstream

### PR DESCRIPTION
**Pull request by nejisama**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Sign the CLA
Make sure you have signed the [CLA](https://www.clahub.com/agreements/alipay/sofa-mosn)

### Solutions

+ Refatcory upstream
   + delete priorityset and some structures that not be used now
   + use hostset to manage subset
   + use types.Cluster instead of simpleInMemCluster struct, makes cluster can be extended.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases.
And you need to show the ut result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/upstream/cluster/host_test.go [hunk 1]
```diff
@@ -1,9 +1,9 @@
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
- * this work for additional information regarding copyright ownership.
- * The ASF licenses this file to You under the Apache License, Version 2.0
- * (the "License"); you may not use this file except in compliance with
+ * tshs work for additional information regarding copyright ownersshp.
+ * The ASF licenses tshs file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use tshs file except in compliance with
  * the License.  You may obtain a copy of the License at
```
Comment (nejisama) [inline pkg/upstream/cluster/host_test.go]: 错误的替换

**Comment by taoyuanyuan** [comment]
Compare init hosts and update hosts performance data

File: pkg/upstream/cluster/cluster_manager.go [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+func (cm *clusterManager) AddOrUpdatePrimaryCluster(cluster v2.Cluster) error {
+	// new cluster
+	newCluster := NewCluster(cluster)
+	if newCluster == nil || reflect.ValueOf(newCluster).IsNil() {
+		return errNilCluster
+	}
+	// check update or new
+	clusterName := cluster.Name
+	pci, loaded := cm.primaryClusterMap.LoadOrStore(clusterName, newPrimaryCluster(newCluster))
+	if loaded { // update
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/cluster_manager.go]: 什么情况下会走入这个逻辑？

File: pkg/upstream/cluster/cluster_manager.go [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+}
+
+// UpdateClusterHosts update all hosts in the cluster
+func (cm *clusterManager) UpdateClusterHosts(clusterName string, hostConfigs []v2.Host) error {
+	pci, ok := cm.primaryClusterMap.Load(clusterName)
+	if !ok {
+		return fmt.Errorf("cluster %s is not exists", clusterName)
+	}
+	pc := pci.(*primaryCluster)
+	var hosts []types.Host
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/cluster_manager.go]: init size with make()

File: pkg/upstream/cluster/cluster_manager.go [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+}
+
+// AppendClusterHosts adds a new host into cluster
+func (cm *clusterManager) AppendClusterHosts(clusterName string, hostConfigs []v2.Host) error {
+	pci, ok := cm.primaryClusterMap.Load(clusterName)
+	if !ok {
+		return fmt.Errorf("cluster %s is not exists", clusterName)
+	}
+	pc := pci.(*primaryCluster)
+	var hosts []types.Host
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/cluster_manager.go]: init size with make()

File: pkg/upstream/cluster/host_set.go [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+	// send refresh notify
+	refreshs := hs.getRefreshNotify()
+	for _, refresh := range refreshs {
+		refresh(host)
+	}
+}
+
+func (hs *hostSet) createSubset(predicate types.HostPredicate) types.HostSet {
+	allHosts := hs.Hosts()
+	var subHosts []types.Host
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/host_set.go]: init size

**Comment by taoyuanyuan** [comment]
```
(pprof) list GetOrCreateAddr
Total: 334511
ROUTINE ======================== sofastack.io/sofa-mosn/pkg/upstream/cluster.GetOrCreateAddr in Go/src/sofastack.io/sofa-mosn/pkg/upstream/cluster/host.go
         0      65537 (flat, cum) 19.59% of Total
         .          .    141:	addr, err := net.ResolveTCPAddr("tcp", addrstr)
         .          .    142:	if err != nil {
         .          .    143:		log.DefaultLogger.Errorf("[upstream] resolve addr %s failed: %v", addrstr, err)
         .          .    144:		return nil
         .          .    145:	}
         .      65537    146:	AddrStore.Store(addrstr, addr)
         .          .    147:	return addr
         .          .    148:}
```

File: pkg/upstream/cluster/loadbalancer.go [hunk 6]
```diff
@@ -27,255 +27,70 @@ import (
...
-	defer l.randMutex.Unlock()
-	idx := l.randInstance.Intn(len(hostSets))
-	hostset := hostSets[idx]
-	hosts := hostset.HealthyHosts()
-	//logger := log.ByContext(context)
-
-	if len(hosts) == 0 {
-		//	logger.Debugf("Choose host failed, no health host found")
+func (lb *randomLoadBalancer) ChooseHost(context types.LoadBalancerContext) types.Host {
+	lb.mutex.Lock()
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/loadbalancer.go]: HealthyHosts()里面也有锁，把lb的lock 放到rand.Intn前面，减少锁力度

File: pkg/upstream/cluster/loadbalancer.go [hunk 7]
```diff
@@ -27,255 +27,70 @@ import (
...
-		return nil
-	}
-
-	selectedHost := selectedHostSet[l.rrIndex%uint32(len(selectedHostSet))]
-	atomic.AddUint32(&l.rrIndex, 1)
-
-	return selectedHost
+func (lb *roundRobinLoadBalancer) ChooseHost(context types.LoadBalancerContext) types.Host {
+	targets := lb.hosts.HealthyHosts()
+	index := atomic.LoadUint32(&lb.rrIndex) % uint32(len(targets))
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/loadbalancer.go]: 并发的时候loadUint32会导致多个chooseHost的index是一样的。
需要修改为 atomic.AddUint32(&lb.rrIndex, 1) % uint32(len(targets)) 类似逻辑

File: pkg/upstream/cluster/host_set.go [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+
+type hostDiff struct {
+	changed      bool
+	finalHosts   []types.Host
+	hostsAdded   []types.Host
+	hostsRemoved []types.Host
+}
+
+func (hs *hostSet) hostsCompare(newHosts, currentHosts []types.Host) *hostDiff {
+	diff := &hostDiff{}
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/host_set.go]: not compare other args， such as tls and meta

**Comment by taoyuanyuan** [comment]
LGTM
